samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -0,0 +1,337 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.execution;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JavaSystemConfig;
+import org.apache.samza.config.JobConfig;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.StreamGraph;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.PartialJoinOperatorSpec;
+import org.apache.samza.system.StreamSpec;
+import org.apache.samza.system.SystemAdmin;
+import org.apache.samza.system.SystemFactory;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamMetadata;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The ExecutionPlanner creates the physical execution graph for the StreamGraph, and
+ * the intermediate topics needed for the execution.
+ */
+public class ExecutionPlanner {
+  private static final Logger log = LoggerFactory.getLogger(ExecutionPlanner.class);
+
+  private final Config config;
+
+  public ExecutionPlanner(Config config) {
+    this.config = config;
+  }
+
+  public ProcessorGraph plan(StreamGraph streamGraph) throws Exception {
+    Map<String, SystemAdmin> sysAdmins = getSystemAdmins(config);
+
+    // create physical processors based on stream graph
+    ProcessorGraph processorGraph = createProcessorGraph(streamGraph);
+
+    if (!processorGraph.getIntermediateStreams().isEmpty()) {
+      // figure out the partitions for internal streams
+      calculatePartitions(streamGraph, processorGraph, sysAdmins);
+
+      // create the streams
+      createStreams(processorGraph, sysAdmins);
+    }
+
+    return processorGraph;
+  }
+
+  /**
+   * Create the physical graph from StreamGraph
+   */
+  /* package private */ ProcessorGraph createProcessorGraph(StreamGraph streamGraph) {
+    // For this phase, we are going to create a processor for the whole dag
+    String processorId = config.get(JobConfig.JOB_NAME()); // only one processor, use the job name
```

Comment:
Good question. I use job name for processorId since it's easier to generate multi-stage job names by appending indexes afterwards, like test-1, test-2. I don't expect multiple instances of the same job exists in the same application so I didn't include both. I agree if we want to support this, we need to have a tuple data structure like ProcessorUid which has both inside.